### PR TITLE
Fix for index error with reading empty parquet file

### PR DESCRIPTION
**Pull request by scharlottej13**
Re #8374, implemented suggested change from @rjzamora to allow stats to be empty.

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the update @scharlottej13! The actual code changes here look good. Would you mind adding a test to ensure this change enables us to read empty parquet files? For instance, we might add something similar to the reproducible example in https://github.com/dask/dask/issues/8374

**Comment by rjzamora** [comment]
>For instance, we might add something similar to the reproducible example in #8374

I agree with this. I was about to suggest a tweak to an existing `test_chunksize_*` test, but they are already quite complicated. A new/simple `test_chunksize_empty` is likely the best bet here :)

**Comment by scharlottej13** [comment]
sounds good, I'll update the PR w/ a simple test, thanks all!

File: dask/dataframe/io/tests/test_parquet.py [hunk 1]
```diff
@@ -2731,6 +2731,20 @@ def test_optimize_and_not(tmpdir):
         assert_eq(a, b)
 
 
+@pytest.mark.parametrize("dtype", ["str", "bool", "float"])
```
Comment (jrbourbeau) [inline dask/dataframe/io/tests/test_parquet.py:2734]: Just wondering what prompted the addition of testing against multiple `dtype`s. Is there some relevant logic for empty DataFrames with different dtypes?

File: dask/dataframe/io/tests/test_parquet.py [hunk 2]
```diff
@@ -2731,6 +2731,20 @@ def test_optimize_and_not(tmpdir):
...
+def test_chunksize_empty(tmpdir, write_engine, read_engine, dtype):
+    df = pd.DataFrame(
+        {
+            "a": pd.Series(dtype="int"),
+            "b": pd.Series(dtype=dtype)
+        }
+    )
+    ddf = dd.from_pandas(df, npartitions=1)
+    ddf.to_parquet(tmpdir, engine=write_engine)
+    dd.read_parquet(tmpdir, engine=read_engine, chunksize="1MiB")
```
Comment (jrbourbeau) [inline dask/dataframe/io/tests/test_parquet.py:2745]: This is nice because it shows that this `dd.read_parquet` now succeeds, while it would fail before. We might also try going one small step further and assert that the DataFrame we get from the `read_parquet` call is the same as the one we started with before the `to_parquet` call. Specifically, we might replace the current `dd.read_parquet(tmpdir, ...)` line with something like:

```python
ddf2 = dd.read_parquet(tmpdir, ...)
assert_eq(ddf, ddf2)
```

**Review by jrbourbeau — COMMENTED** [review]
Thanks for adding a test @scharlottej13. I left a couple of small comments, but otherwise this looks good to me

Additionally, it looks like the code linting checks in CI are failing. You can run those linting checks locally using the instructions here https://docs.dask.org/en/stable/develop.html#code-formatting

File: dask/dataframe/io/tests/test_parquet.py (hunk 1 above)
Comment (scharlottej13) [reply dask/dataframe/io/tests/test_parquet.py:2734]: Good question! I think I was conflating this issue with #8373, I can take this part out

File: dask/dataframe/io/tests/test_parquet.py (hunk 2 above)
Comment (scharlottej13) [reply dask/dataframe/io/tests/test_parquet.py:2745]: oh great! I was wondering if an assertion like this would be helpful, I'll add it in

**Comment by scharlottej13** [comment]
After adding an `assert_eq` to the test, it now fails when writing with `pyarrow` and reading with `fastparquet` (but not the other way round). [Here](https://github.com/dask/dask/runs/4302880979?check_suite_focus=true#step:6:21707) is an example of one of the failed tests and the specific error is [here](https://github.com/dask/dask/runs/4302880979?check_suite_focus=true#step:6:21128). The data types of the indices were also different, which is why I set `check_index=False` in the test.

Doing this without Dask, the data types still seem to be different:
```python
import pandas as pd

df = pd.DataFrame({"a": pd.Series(dtype="int"), "b": pd.Series(dtype="str")})
df.to_parquet("~/tmp/x.parq", engine="pyarrow")
df2 = pd.read_parquet("~/tmp/x.parq", engine="fastparquet")

df.dtypes
Out[5]: 
a     int64
b    object
dtype: object

df2.dtypes
Out[6]: 
a    float64
b    float64
dtype: object
```
python 3.9.7
pandas 1.3.4
fastparquet 0.6.3 (and @ian-r-rose also tested w/ the newest version)

@martindurant maybe you have some insight into this?

**Comment by martindurant** [comment]
The parquet schema that pyarrow stores looks like
```
- schema: REQUIRED
| - a: INT64, OPTIONAL
| - b: INT32, UNKNOWN, OPTIONAL
  - __index_level_0__: INT32, UNKNOWN, OPTIONAL
```
(where "UNKNOWN" is the logical NULL type)
So I guess that pyarrow is doing a special case here and dropping the information about strings.

Note that in the current version of fastparquet, the columns come back as int types, not float. I don't remember, but I assume there was a bug for when the number of NULL values was not known.

**Comment by ian-r-rose** [comment]
Thanks @martindurant, a small reproducer extending what @scharlottej13 posted but is independent of fastparquet:

```python
import pandas as pd
import pyarrow.parquet as pq

df = pd.DataFrame({"a": pd.Series([1], dtype="int"), "b": pd.Series(["c"], dtype="str")})
df.to_parquet("~/tmp/x.parq", engine="pyarrow")
pq.read_schema("/home/ian/tmp/x.parq")
```
produces
```
a: int64
  -- field metadata --
  PARQUET:field_id: '1'
b: null
  -- field metadata --
  PARQUET:field_id: '2'
-- schema metadata --
pandas: '{"index_columns": [{"kind": "range", "name": null, "start": 0, "' + 457
```

Perhaps @jorisvandenbossche knows why the string dtype information is getting lost?

For this PR, I'd suggest changing the test to involve a different dtype (e.g., `float`), and opening a new issue for the `str` round-trip where appropriate (once we determine that).

**Review by jrbourbeau — APPROVED** [review]
+1 for @ian-r-rose's suggestion to include the improvement here as is and then open up a separate ticket for the `str` round-trip issue which was uncovered. @scharlottej13 would you be comfortable opening up that issue? 

Thanks for sticking with this PR @scharlottej13! This is in 🎉 

Also, I noticed this is your first code contribution to this repository. Welcome!
